### PR TITLE
Fix lockup when filtering asset browser entries by type

### PR DESCRIPTION
**Pull request by gadams3**
## What does this PR do?

Optimize asset browser entry filters to use visitors instead of recursively building child containers
Optimize asset browser entry filters to use sets instead of vectors
Optimize asset browser entry filter string comparisons Optimize asset browser table view
Added expand all button to asset browser menu

Fixes https://github.com/o3de/o3de/issues/17091
Fixes https://github.com/o3de/o3de/issues/14645
Fixes https://github.com/o3de/o3de/issues/17247
Fixes https://github.com/o3de/o3de/issues/16272
 Fixes https://github.com/o3de/o3de/issues/14645

## How was this PR tested?

Tested performance of asset browser filters using the multiplayer sample project
Enabling and disabling asset browser type filters updates immediately instead of taking several seconds
This should also improve performance of other asset browser operations that utilize filters

### CODE REVIEW

**Review by AMZN-daimini — APPROVED** [review]

**Review by AMZN-alexpete — APPROVED** [review]
Wow so many fixes and optimizations, amazing!

File: Code/Framework/AzToolsFramework/AzToolsFramework/AssetBrowser/Entries/AssetBrowserEntry.h [hunk 1]
```diff
@@ -137,6 +141,14 @@ namespace AzToolsFramework
             const QString& GetEntryTypeAsString() const;
             static const AZStd::string ExtensionToType(AZStd::string_view str);
 
+            //! Call the visitor function for the current entry and all of its parents.
```
Comment (spham-amzn) [inline Code/Framework/AzToolsFramework/AzToolsFramework/AssetBrowser/Entries/AssetBrowserEntry.h:144]:
```suggestion
            //! Recursively call the visitor function for the current entry and all of its parents.
```

File: Gems/Atom/RPI/Code/Source/RPI.Builders/Material/MaterialTypeBuilder.cpp [hunk 2]
```diff
@@ -493,7 +493,8 @@ namespace AZ
                 }
 
                 // Intermediate azsl file
-                AZStd::string generatedAzsl = AZStd::string::format("// This code was generated by %s. Do not modify.\n\n", MaterialTypeBuilderName);
+                AZStd::string generatedAzsl;
+                generatedAzsl += AZStd::string::format("// This code was generated by %s. Do not modify.\n", MaterialTypeBuilderName);
```
Comment (spham-amzn) [inline Gems/Atom/RPI/Code/Source/RPI.Builders/Material/MaterialTypeBuilder.cpp:497]: Why was this moved from being initialized in the constructor?

**Review by spham-amzn — APPROVED** [review]

File: Gems/Atom/RPI/Code/Source/RPI.Builders/Material/MaterialTypeBuilder.cpp (hunk 2 above)
Comment (gadams3) [reply Gems/Atom/RPI/Code/Source/RPI.Builders/Material/MaterialTypeBuilder.cpp]: There were different versions of the file when I merged. One version have the lines split like this further down in the file. That's the version I kept. History of changes was lost when I retargeted to development branch.

File: Code/Framework/AzToolsFramework/AzToolsFramework/AssetBrowser/Search/Filter.h [hunk 3]
```diff
@@ -52,8 +50,12 @@ namespace AzToolsFramework
                 Down    = 0x02
             };
 
-            AssetBrowserEntryFilter();
-            virtual ~AssetBrowserEntryFilter() =  default;
+            AssetBrowserEntryFilter() = default;
+            ~AssetBrowserEntryFilter() override = default;
+
+            //! Cloning function that must be overridden for certain asset browser views that duplicate and modify incoming filters\
```
Comment (gadams3) [inline Code/Framework/AzToolsFramework/AzToolsFramework/AssetBrowser/Search/Filter.h:56]:
```suggestion
            //! Cloning function that must be overridden for certain asset browser views that duplicate and modify incoming filters
```
